src/System.CommandLine/Parser.cs

```diff
@@ -30,7 +30,7 @@ public Parser() : this(new RootCommand())
         {
             var rawTokens = arguments;  // allow a more user-friendly name for callers of Parse
             var lexResult = NormalizeRootCommand(rawTokens).Lex(Configuration);
-            var directives = new List<string>();
+            var directives = new Dictionary<string, string>();
```

Comment:
I don't think we want to use a `Dictionary` here. The order of items when enumerating a dictionary is explicitly undefined, so exposing them in this way causes the order to be lost. Perhaps keep it as a `List<KeyValue<,>>`?

Reply:
Interesting point. I realize that my assumption is that order should not be significant for directives. Did you have a use case in mind for preserving order?

(Related, I think that the token collections on the `ParseResult` do preserve order, though I'm not sure this is tested.)

Reply:
I am not sure that there is a use case with the order current directives in core. I am thinking of if someone puts in their own directives that have an ordering dependency. This is certainly an edge case.

Reply:
Agreed. I am less concerned about the middleware, are more just noting that anyone manually processing the `ParseResult` will (to some degree since as you point out your could go back to the Tokens to look up order) lose access to the order they were specified.

Reply:
I agree with Kevin that the order could matter if somebody adds their own directives. 

@jonsequitur What was your thought on the value the dictionary provides? The directives are parsed once, so I don't see a time trade off. `OrderdDictionary` could work, if there is another reason. Duplicate directives that contradict each other (e.g. render=ansi and render=notthat) would be processed twice in a collection, but in general I don't think we need to handle duplication.

Reply:
@jeredm Thanks! I just realized duplicates aren't tested, and will result in an exception with this change.

Reply:
It's come up a few times that directives with values would be useful, e.g. for render mode hints. It makes sense for the parsing rules for these values to be consistent across directives. The dictionary was the simplest choice. 

With a dictionary, one can test for presence of a directive independently from a value, if the handler cares about the value. Maybe a custom type would be better? I opted for strings because binding to other types get us into much more complex territory (validation, arity, etc.), although the argument could be made that directives should be derived from `ISymbol` and support all of this as well as completions.